builtin/game/falling.lua

```diff
@@ -56,78 +48,91 @@ core.register_entity(":__builtin:falling_node", {
-			-- If it's not air or liquid, remove node and replace it with
-			-- it's drops
-			if n2.name ~= "air" and (not nd or nd.liquidtype == "none") then
-				core.remove_node(np)
-				if nd and nd.buildable_to == false then
-					-- Add dropped items
-					local drops = core.get_node_drops(n2, "")
+		end
+
+		-- Stop node if it falls on walkable surface, or floats on water
+		if (below_nodef and below_nodef.walkable == true)
+		or (below_nodef
+				and core.get_item_group(self.node.name, "float") ~= 0
+				and below_nodef.liquidtype ~= "none") then
+
+			self.object:set_velocity({x = 0, y = 0, z = 0})
+		end
+
+		-- Has the fallen node stopped moving ?
+		local vel = self.object:get_velocity()
+		if vector.equals(vel, {x = 0, y = 0, z = 0}) then
+			local npos = self.object:get_pos()
+			-- Get node we've landed inside
+			local cnode = minetest.get_node(npos).name
+			local cdef = core.registered_nodes[cnode]
+			-- If 'air' or buildable_to or an attached_node then place node,
+			-- otherwise drop falling node as an item instead.
+			if cnode == "air"
+			or (cdef and cdef.buildable_to == true)
+			or (cdef and cdef.liquidtype ~= "none")
+			or core.get_item_group(cnode, "attached_node") ~= 0 then
+				-- Are we an attached node ? (grass, flowers, torch)
+				if core.get_item_group(cnode, "attached_node") ~= 0 then
+					-- Add drops from attached node
+					local drops = core.get_node_drops(cnode, "")
 					for _, dropped_item in pairs(drops) do
-						core.add_item(np, dropped_item)
+						core.add_item(npos, dropped_item)
+					end
+					-- Run script hook
+					for _, callback in pairs(core.registered_on_dignodes) do
+						callback(npos, cnode)
 					end
 				end
-				-- Run script hook
-				for _, callback in pairs(core.registered_on_dignodes) do
-					callback(np, n2)
-				end
-			end
-			-- Create node and remove entity
-			local def = core.registered_nodes[self.node.name]
-			if def then
-				core.add_node(np, self.node)
+				-- Round position
+				npos = vector.round(npos)
+				-- Place falling entity as node and write any metadata
+				core.add_node(npos, self.node)
 				if self.meta then
-					local meta = core.get_meta(np)
+					local meta = core.get_meta(npos)
 					meta:from_table(self.meta)
 				end
+				-- Play placed sound
+				local def = core.registered_nodes[self.node.name]
 				if def.sounds and def.sounds.place and def.sounds.place.name then
-					core.sound_play(def.sounds.place, {pos = np})
+					core.sound_play(def.sounds.place, {pos = npos})
+				end
+				-- Just incase we landed on other falling nodes
+				core.check_for_falling(npos)
+			else
+				-- Add drops from falling node
+				local drops = core.get_node_drops(self.node, "")
+				for _, dropped_item in pairs(drops) do
+					core.add_item(npos, dropped_item)
 				end
 			end
+			-- Remove falling entity
 			self.object:remove()
-			core.check_for_falling(np)
-			return
-		end
-		local vel = self.object:get_velocity()
```

Comment:
This was a bugfix where mods added horizontal velocity to falling nodes. It can happen that the falling node is stuck on the corner of another node, but the node below its centre is air.